### PR TITLE
[Enhancement] Support multi-part GRF generated by COLOCATE/SHUFFLE_HASH_BUCKET HJ with only one eqConjunct

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/6372
## Problem Summary(Required) ：

COLOCATE/SHUFFLE_HASH_BUCKET HJ does not produce global runtime filters, this PR support these HJ with only one eqConjunct generated a shuffle-aware(multipart) GRF. In the future, partition_by_exprs will be introduced to support multipart GRF generated by HJ with mutiple eqConjuncts.

When evaluating a GRF generated by COLOCATE HJ; at first,  crc32_hash is used to compute a hash_value, then it is modulo-ed by num_buckets of colocate group to compute the bucket_seq, a bucket_seq to fragment instance ordinal mapping is used to translate bucket_seq to partition index of multipart GRF, then the part of GRF indexed at partition index is probed.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 16 / 43 (37.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|8|31|25.81%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|7|11|63.64%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|1|1|100.00%|

File: be/src/exprs/vectorized/runtime_filter.h [hunk 1]
```diff
@@ -336,13 +337,51 @@ class RuntimeBloomFilter final : public JoinRuntimeFilter {
...
+            hash_values.assign(num_rows, HashUtil::FNV_SEED);
+            (input_column->*hash_func)(hash_values.data(), 0, num_rows);
+            for (size_t i = 0; i < num_rows; i++) {
+                hash_values[i] %= num_hash_partitions;
+            }
+        };
+
+        switch (join_mode) {
+        case TRuntimeFilterBuildJoinMode::BORADCAST: {
+            hash_values.assign(num_rows, 0);
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/runtime_filter.h:362]: Why need to assign hash value for boradcast join?

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 2]
```diff
@@ -1333,6 +1337,26 @@ private void computeFragmentHosts() throws Exception {
...
     }
 
+    public void computeBucketSeq2InstanceOrdinal(FragmentExecParams params, boolean hasColocate) {
+        if (!hasColocate) {
+            return;
+        }
+        Integer bucketNum = fragmentIdToBucketNumMap.get(params.fragment.getFragmentId());
+        Preconditions.checkArgument(bucketNum != null, "bucketNum not exists for colocate Fragment");
+        Integer[] bucketSeq2InstanceOrdinal = new Integer[bucketNum];
+        for (int bucketSeq = 0; bucketSeq < bucketNum; ++bucketSeq) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1347]: needn't assign -1.

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 3]
```diff
@@ -213,6 +214,8 @@ public class Coordinator {
     // fragment_id -> < be_id -> bucket_count >
     private final Map<PlanFragmentId, Map<Long, Integer>> fragmentIdToBackendIdBucketCountMap = Maps.newHashMap();
 
+    private final Map<PlanFragmentId, List<Integer>> fragmentIdToSeqToInstanceMap = Maps.newConcurrentMap();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:217]: needn't concurrent.

File: be/src/exprs/vectorized/runtime_filter.h (hunk 1 above)
Comment (satanson) [reply be/src/exprs/vectorized/runtime_filter.h:362]: keep consistent with the old code. delete it ?

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java (hunk 2 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1347]: ok

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java (hunk 3 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:217]: yes

File: be/src/exprs/vectorized/runtime_filter.h [hunk 4]
```diff
@@ -336,13 +337,51 @@ class RuntimeBloomFilter final : public JoinRuntimeFilter {
...
         }
     }
 
+    void compute_hash_values_for_multi_part(RunningContext* running_ctx, int8_t join_mode, const Column* input_column,
+                                            std::vector<uint32_t>& hash_values, size_t num_rows) const {
+        typedef void (Column::*HashFuncType)(uint32_t*, uint32_t, uint32_t) const;
+
+        auto compute_hash = [&input_column, &num_rows, &hash_values, this](HashFuncType hash_func,
+                                                                           size_t num_hash_partitions) {
+            hash_values.assign(num_rows, HashUtil::FNV_SEED);
```
Comment (ZiheLiu) [inline be/src/exprs/vectorized/runtime_filter.h:353]: Init value of `fnv_hash` is `HashUtil::FNV_SEED` , but init value of `crc32_hash` is 0.

Comment (satanson) [reply be/src/exprs/vectorized/runtime_filter.h:353]: yes

File: be/src/exprs/vectorized/runtime_filter.h [hunk 5]
```diff
@@ -336,13 +337,51 @@ class RuntimeBloomFilter final : public JoinRuntimeFilter {
...
+            break;
+        }
+        case TRuntimeFilterBuildJoinMode::LOCAL_HASH_BUCKET: {
+            compute_hash(&Column::crc32_hash, _num_hash_partitions);
+            break;
+        }
+        case TRuntimeFilterBuildJoinMode::COLOCATE: {
+            compute_hash(&Column::crc32_hash, running_ctx->bucketseq_to_partition.size());
+            for (auto i = 0; i < num_rows; ++i) {
+                hash_values[i] = running_ctx->bucketseq_to_partition[hash_values[i]];
```
[the remark this answers is not in the document]
Comment (satanson) [reply be/src/exprs/vectorized/runtime_filter.h:382]: commented

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 39 / 44 (88.64%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|24|28|85.71%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|15|16|93.75%|

File: gensrc/thrift/RuntimeFilter.thrift [hunk 6]
```diff
@@ -29,8 +29,10 @@ enum TRuntimeFilterBuildJoinMode {
   NONE,
   BORADCAST,
   PARTITIONED,
-  BUCKET_SHUFFLE,
-  COLOCATE
+  LOCAL_HASH_BUCKET,
```
Comment (stdpain) [inline gensrc/thrift/RuntimeFilter.thrift:32]: Is there a grayscale upgrade problem here ?

Comment (satanson) [reply gensrc/thrift/RuntimeFilter.thrift:32]: it is ok, just change the symbolic names.

**Review by stdpain — DISMISSED** [review]

File: fe/fe-core/src/test/resources/sql/tpchcost/q7.sql [hunk 7]
```diff
@@ -116,7 +116,7 @@ OutPut Exchange Id: 22
 |  equal join conjunct: [39: C_NATIONKEY, INT, false] = [50: N_NATIONKEY, INT, false]
 |  other join predicates: ((46: N_NAME = 'CANADA') AND (51: N_NAME = 'IRAN')) OR ((46: N_NAME = 'IRAN') AND (51: N_NAME = 'CANADA'))
 |  build runtime filters:
-|  - filter_id = 3, build_expr = (50: N_NATIONKEY), remote = true
+|  - filter_id = 3, build_expr = (50: N_NATIONKEY), remote = false
```
Comment (kangkaisen) [inline fe/fe-core/src/test/resources/sql/tpchcost/q7.sql:119]: Why remote change from true to false?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 70 / 76 (92.11%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|24|28|85.71%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|15|16|93.75%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|31|32|96.88%|

File: fe/fe-core/src/test/resources/sql/tpchcost/q7.sql (hunk 7 above)
Comment (satanson) [reply fe/fe-core/src/test/resources/sql/tpchcost/q7.sql:119]: it is something wrong happens. I will check

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 70 / 79 (88.61%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|24|28|85.71%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|31|35|88.57%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|15|16|93.75%|

**Review by ZiheLiu — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

**Comment by satanson** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 70 / 79 (88.61%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|24|28|85.71%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|31|35|88.57%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|15|16|93.75%|

**Review by kangkaisen — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Pull request merged**